Tutorial: Cross-chain NFT marketplace (#1634)
iotaledger/iota-wiki: docs/tutorials/cross-chain-nft-marketplace-part-1.md

![alt text](../../static/img/tutorials/cross_chain_marketplace/Architecture-V2.png)

Finally, After deploying another instance of the marketplace on the BNB Testnet in Part III, where the contract will handle cross-chain transactions, the architecture will look like this:
![alt text](../../static/img/tutorials/cross_chain_marketplace/Architecture-V3.png)

This enables a user, e.g on BNB Testnet, to view and buy an NFT listed on the ShimmerEVM Testnet and vice versa without needing to switch networks.

## Prerequisites

- [Node.js v18](https://hardhat.org/tutorial/setting-up-the-environment#installing-node.js) and above supported.

- [npx](https://www.npmjs.com/package/npx) v7.1.0 and above supported.

## Set Up

clone the [tutorial repository](https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace) and navigate to the project folder, then run:

```bash
npm install
```

## Contracts

For the scope of this part, we wil need two contracts: an NFTMarketplace contract, and an NFT ERC721-compatible contract.

> Create a `contracts` folder in the root of the project and add the following files under it:

Reviewer: It'd be good to explain the contracts as well, you can use reference imports with line numbers to import only the relevant section. Let me know if you need any help with this.

run it by executing:

```bash
npx hardhat run scripts/deploy_marketplace_shimmer.js --network shimmerevm-testnet
```

### deploy_er721_shimmer.js

```javascript
const fs = require('fs');
const path = require('path');

async function main() {
    const MyERC721 = await ethers.getContractFactory("MyERC721");
    const myERC721 = await MyERC721.deploy("SampleToken", "SESA", "SampleTokenURI");

    const myERC721Address = await myERC721.getAddress();

    console.log("MyERC721 deployed to:", myERC721Address);

    const addressDirectory = path.join(__dirname, 'addresses');
    fs.mkdirSync(addressDirectory, { recursive: true }); // Ensure the directory exists, create it if it doesn't

    const filePath = path.join(addressDirectory, 'MyERC721.txt');
    fs.writeFileSync(filePath, myERC721Address);

    console.log(`Contract address written to ${filePath}`);
}

main()
    .then(() => process.exit(0))
    .catch((error) => {
        console.error(error);
        process.exit(1);
    });
```
This will deploy the MyERC721 contract to the ShimmerEVM Testnet and save the contract address to a file.

> run it by executing:

Reviewer: ```suggestion
You can run the script by executing the following command:
```

```bash
npx hardhat run scripts/deploy_er721_shimmer.js --network shimmerevm-testnet
```

### mint_nft.js

> After deploying the MyERC721 contract, let's mint an NFT by running the following script:

Reviewer: ```suggestion
After you have deployed the `MyERC721` contract, you are ready to mint an NFT using the following script:
```

> ```javascript
> const fs = require('fs');
> const path = require('path');
> 
> async function createNFT(myERC721Address) {
> 
>     const MyERC721 = await ethers.getContractFactory("MyERC721");
>     const myERC721 = MyERC721.attach(myERC721Address);
> 
>     const tx = await myERC721.mint();
>     await tx.wait(); // Wait for the transaction to be mined
> }
> 
> async function main() {
>     // Read the contract address from the file
>     const addressPath = path.join(__dirname, 'addresses', 'MyERC721.txt');
>     const myERC721Address = fs.readFileSync(addressPath, 'utf8').trim();
> 
>     await createNFT(myERC721Address);
> }
> 
> main()
>     .then(() => process.exit(0))
>     .catch((error) => {
>         console.error(error);
>         process.exit(1);
>     });
> ```

Reviewer: ````suggestion
```javascript reference
https://github.com/iota-community/ISC-Cross-Chain-NFT-Marketplace/blob/main/scripts/mint_nft.js
```
````

> execute the script by running:

Reviewer: ```suggestion

You can run the script by executing the following command:
```

```bash
npx hardhat run scripts/mint_nft.js --network shimmerevm-testnet
```

> 
> 
> ### approve_myERC721_for_marketplace.js

Reviewer: ```suggestion
### approve_myERC721_for_marketplace.js
```
